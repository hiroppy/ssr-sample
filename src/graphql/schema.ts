import { makeExecutableSchema, IResolvers } from 'graphql-tools';
import { apolloSamples } from '../server/responseSchema';

export const typeDefs = `
  type Sample {
    id: Int!
    code: String!
    name: String!
    likeCount: Int!
    description: String
  }

  type Query {
    samples(maxLength: Int): [Sample]
  }

  type Mutation {
    addLike(id: Int): Sample
  }
`;

export const resolvers: IResolvers = {
  Query: {
    samples: (obj, { maxLength }: { maxLength?: number }) => {
      return apolloSamples.slice(0, maxLength || apolloSamples.length);
    }
  },
  Mutation: {
    addLike: (obj, { id }: { id: number }) => {
      const item = apolloSamples.find(({ id: fileId }) => id === fileId);

      if (item) {
        ++item.likeCount;

        return item;
      } else {
        throw new Error('Could not find');
      }
    }
  }
};

export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});
