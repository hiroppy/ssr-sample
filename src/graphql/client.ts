import { ApolloClient, InMemoryCache, ApolloLink } from 'apollo-boost';

// link is an argument because schema.ts used by SchemaLink uses `fs` module
export function createClient({ link }: { link: ApolloLink }) {
  return new ApolloClient({
    ssrMode: !process.env.IS_BROWSER,
    link,
    cache: process.env.IS_BROWSER
      ? new InMemoryCache().restore((window as any).__APOLLO_STATE__)
      : new InMemoryCache()
  });
}
