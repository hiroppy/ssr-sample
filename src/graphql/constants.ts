export const endpoint = '/graphql';
export const limitCost = 10;
