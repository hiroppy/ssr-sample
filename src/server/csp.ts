import { Request, Response, NextFunction } from 'express';
import * as helmet from 'helmet';
import nanoid from 'nanoid';
import ua from 'ua-parser-js';

export function generateNonceId(req: Request, res: Response, next: NextFunction) {
  res.locals.nonce = Buffer.from(nanoid(32)).toString('base64');
  next();
}

// google font: https://stackoverflow.com/a/34576000/7014700
const baseDirectives: helmet.IHelmetContentSecurityPolicyDirectives = {
  defaultSrc: ["'self'"],
  styleSrc: ["'unsafe-inline'", 'fonts.googleapis.com'], // for styled-components
  fontSrc: ["'self'", 'data: fonts.gstatic.com'],
  imgSrc: ["'self'", 'img.shields.io'], // for README
  connectSrc: ["'self'", 'img.shields.io', 'fonts.googleapis.com', 'fonts.gstatic.com'], // for service-worker
  workerSrc: ["'self'"]
};

// chrome, firefox
const lv3Directives: helmet.IHelmetContentSecurityPolicyDirectives = {
  ...baseDirectives,
  scriptSrc: [(req, res) => `'nonce-${res.locals.nonce}'`, "'strict-dynamic'", "'unsafe-eval'"]
};

// safari
// TODO: cannot load a file to be imported as dynamic because cannot use strict-dynamic
const lv2Directives: helmet.IHelmetContentSecurityPolicyDirectives = {
  ...baseDirectives,
  scriptSrc: [
    "'self",
    (req, res) => `'nonce-${res.locals.nonce}'`,
    "'unsafe-eval'",
    "'unsafe-inline'"
  ]
};

export function csp(req: Request, res: Response, next: NextFunction) {
  const directives = ['Chrome', 'Firefox'].includes(
    // TODO: fix
    (ua as any)(req.headers['user-agent']).browser.name
  )
    ? lv3Directives
    : lv2Directives;

  helmet.contentSecurityPolicy({ directives })(req, res, next);
}
