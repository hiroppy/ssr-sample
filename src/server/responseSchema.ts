import raw from 'raw.macro';

export type Samples = typeof apolloSamples;

// export type Themes = {
//   [key in 'dark' | 'light']: {
//     main: string;
//     sub: string;
//   };
// };

// TODO: fix
// need to create babel-plugin

export const sagaSamples = [
  {
    id: 1,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];

export const apolloSamples = [
  {
    id: 1,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];
