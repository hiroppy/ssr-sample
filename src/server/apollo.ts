import * as express from 'express';
import { ApolloServer } from 'apollo-server-express';
import { getComplexity, simpleEstimator, fieldExtensionsEstimator } from 'graphql-query-complexity';
import { separateOperations } from 'graphql';
import { resolvers, schema } from '../graphql/schema';
import { endpoint, limitCost } from '../graphql/constants';

export function apollo(app: express.Application) {
  const apollo = new ApolloServer({
    debug: process.env.NODE_ENV !== 'production',
    schema,
    resolvers,
    playground:
      process.env.NODE_ENV !== 'production'
        ? {
            endpoint
          }
        : false,
    plugins: [
      {
        requestDidStart: () => ({
          didResolveOperation({ request, document }) {
            const complexity = getComplexity({
              schema,
              query: request.operationName
                ? separateOperations(document)[request.operationName]
                : document,
              variables: request.variables,
              estimators: [fieldExtensionsEstimator(), simpleEstimator({ defaultComplexity: 1 })]
            });

            if (complexity >= limitCost) {
              throw new Error(`${complexity} is over ${limitCost}`);
            }
            console.log('Used query complexity points:', complexity);
          },
          didEncounterErrors({ errors, request }) {
            // e.g. add a logger
            console.error(errors);
            console.error(request);
          }
        })
      }
    ]
  });

  apollo.applyMiddleware({ app, path: endpoint });
}
