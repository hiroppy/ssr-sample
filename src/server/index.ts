import { join } from 'path';
import cluster from 'cluster';
import { cpus } from 'os';
import { config } from 'dotenv';
import { runServer } from './server';

const isProd = process.env.NODE_ENV === 'production';

// If you compile server code with webpack, this is unnecessary.
config({
  path: isProd ? join(__dirname, '../../../.env.prod') : join(__dirname, '../../.env.dev')
});

if (isProd) {
  const numCPUs = cpus().length;

  if (cluster.isMaster) {
    [...new Array(numCPUs)].forEach(() => cluster.fork());

    // cluster manager
    cluster.on('exit', (worker, code, signal) => {
      console.log(`Restarting ${worker.process.pid}. ${code || signal}`);
      cluster.fork();
    });
  } else {
    runServer();
  }
} else {
  runServer();
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

process.on('unhandledRejection', (err) => {
  console.error(err);
});
