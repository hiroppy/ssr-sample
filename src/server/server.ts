/* eslint @typescript-eslint/no-var-requires: 0 */

import { createServer } from 'http';
import express from 'express';
import * as bodyParser from 'body-parser';
import compression from 'compression';
import { generateNonceId, csp } from './csp';
import { apollo } from './apollo';
import { router } from './router';

export function runServer() {
  const app = express();
  const port = process.env.PORT || 3000;

  app.use(bodyParser.urlencoded({ extended: true }));
  app.use(bodyParser.json());

  // CSP
  app.use(generateNonceId);
  app.use(csp);

  // compression
  app.use(compression({ level: 5 }));

  // HMR
  if (process.env.NODE_ENV === 'development') {
    const webpack = require('webpack');
    const webpackHotMiddleware = require('webpack-hot-middleware');
    const webpackDevMiddleware = require('webpack-dev-middleware');
    const config = require('../../webpack.client.config');
    const compiler = webpack(config);

    app.use(webpackHotMiddleware(compiler));
    app.use(
      webpackDevMiddleware(compiler, {
        publicPath: config.output.publicPath,
        writeToDisk(filePath: string) {
          return /loadable-stats/.test(filePath);
        }
      })
    );
  }

  // setup apollo
  apollo(app);

  // register routes
  router(app);

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'test') {
    const server = createServer(app);

    server.listen(port, () => {
      console.log(`Listening on ${port}`);
    });

    // https://stackoverflow.com/a/48710483/7014700
    server.on('error', (err: NodeJS.ErrnoException) => {
      if (err.syscall !== 'listen') throw err;

      const bind = typeof port === 'string' ? `Pipe ${port}` : `Port ${port}`;

      switch (err.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw err;
      }
    });
  }

  return app;
}
