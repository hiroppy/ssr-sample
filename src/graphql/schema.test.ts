import { graphql } from 'graphql';
import { schema } from './schema';
import { apolloSamples } from '../server/responseSchema';

test('should return result from samples resolver', async () => {
  const GET_SAMPLES = `
    query getSamples($maxLength: Int) {
      samples(maxLength: $maxLength) {
        id
        name
        code
        likeCount
        description
      }
    }
  `;

  const res = await graphql(
    schema,
    GET_SAMPLES,
    null,
    {
      samples: apolloSamples
    },
    {
      maxLength: 1
    }
  );

  expect(res).toMatchSnapshot();
});

test('should return result from addLike resolver', async () => {
  const ADD_LIKE = `
    mutation addLike($id: Int) {
      addLike(id: $id) {
        id
      }
    }
  `;

  {
    const res = await graphql(
      schema,
      ADD_LIKE,
      null,
      {
        samples: apolloSamples
      },
      {
        id: 1
      }
    );

    expect(res).toMatchSnapshot();
  }

  {
    const res = await graphql(
      schema,
      ADD_LIKE,
      null,
      {
        samples: apolloSamples
      },
      {
        id: 1000
      }
    );

    expect(res).toMatchSnapshot();
  }
});
