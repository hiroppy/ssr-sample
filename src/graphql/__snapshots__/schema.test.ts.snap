// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return result from addLike resolver 1`] = `
Object {
  "data": Object {
    "addLike": Object {
      "id": 1,
    },
  },
}
`;

exports[`should return result from addLike resolver 2`] = `
Object {
  "data": Object {
    "addLike": null,
  },
  "errors": Array [
    [GraphQLError: Could not find],
  ],
}
`;

exports[`should return result from samples resolver 1`] = `
Object {
  "data": Object {
    "samples": Array [
      Object {
        "code": "test-code",
        "description": "aaaa",
        "id": 1,
        "likeCount": 0,
        "name": "apollo.tsx",
      },
    ],
  },
}
`;
